src/connector/src/sink/encoder/mod.rs

```diff
@@ -82,6 +82,16 @@ impl<T> SerTo<T> for T {
     }
 }
 
+pub trait QueryField {
```

Comment:
Instead of querying the topic field from the formatted payload, it's better to get the field directly from the stream chunk before formatting it. This can be implemented in the `write_chunk` method.

A `StreamChunk` can be seen as `Vec<(Op, OwnedRow)>`, where `Op` is row operation like `Insert` `Delete`, and `OwnedRow` is the rows in the stream chunk with the same schema as the sink schema. You can get the topic field string directly from the row with the index of the field column.

In sink validation, you can check the sink schema to ensure that the topic field is of String type. When building the sink writer, you save the index of the topic field column if there is any, to avoid repeatedly querying with field name string.

Reply:
@wenym1 That was my initial approach, but the issue I found is that there are no guarantees about the ordering of the rows and the returned payload (what happens if at any point we execute the formatting in parallel) nor a way to map them (since there're no primary keys in all the cases).  Do you think the order of the rows will always be the same as the formatted chunks if that's the case then I guess the Interface can be removed.

I'll make the changes about the validation and using the default topic as a fallback, through a property

Reply:
First of all, I can clarify the difference between formatter and encoder. Formatter specifies the way we handle a series of rows with op. For example, the append only formatter drops all row with op that is not `Insert`, while debezium formatter merge the row with op `UpdateDelete` and `UpdateInsert` into a single change log message. Encoder specifies the way we convert a `Row` into output payload, such as json encoder. 

For mqtt sink, IIUC, there is no primary keys, and it can only handle the append only workload, and therefore only append-only formatter will be used. For append only formatter, if the input workload is append only, you can expect that the returned payload will have the same row order and same row count as the input. 

Actually, for simplicity, since only the trivial append only formatter is used, I would recommend you use encoder only and get rid of the formatter, and then you don't have to worry about the implicit behavior of formatter. You can manually iter over all rows in the stream chunk, fetch the topic, and encode the payload, and then write to the external mqtt.

Reply:
@wenym1  Having the encoder only makes a lot of sense in the `mqtt` sink, I still think these could be beneficial for other non-append-only sinks, but it's not part of the scope of this PR

Reply:
If we want to generalize the functionality that gets the topic field from the row, I think it's still not a good idea to get the topic from the encoded payload, and it's more intuitive and elegant to get the topic from the row directly. 

Getting topic from the row is something like getting an output from the input row, and it works like an `RowEncoder` that encodes an input row into an output topic string. To generalize this functionality, you can try implementing it as a wrapper row encoder that wraps an inner encoder. Its output type will be (string, <output of inner encoder>). This encoder can be embedded to the generic formatters (e.g. AppendOnlyFormatter) we have implemented so that in the future we can extend and reuse it on other sinks if needed.